test/tests/wa_has_one_optimization_test.rb

```diff
@@ -0,0 +1,151 @@
+# frozen_string_literal: true
+
+require_relative "../test_helper"
+
+# has_one requires doing sub-queries with SELECT IN / applying limits to
+# handle conditions / counts properly. However, when there is no conditions and counts,
+# the deepest layer can avoid those sub-queries, which is faster.
+
+module RegexpTricks
+  # Annoyingly, some older versions of rails sometimes have extra spacing...
+  def self.regex_for_any_spaces(string)
```

Comment:
Nice trick. I wonder if it would be clearer to squeeze multiple spaces into one in the actual SQL and compare against the expected canonical SQL with single spaces. Something like:
```ruby
    sql = S0.where_assoc_exists(:o1).to_sql.squeeze(" ") # perhaps in a helper method
    assert_match 'SELECT 1 FROM "s1s"', sql
    refute_match 'SELECT 1 FROM (SELECT "s1s".*', sql
```

Reply:
Great idea! Avoiding regex whenever a clean alternative is available is always a good idea!